### DISCUSSION TITLE
How to handle file upload on client using react hook form and next.js 14 server actions

### DISCUSSION THREAD

**Discussion by lyubk0** [vercel/next.js#68944] [Help] [answered]
### Summary

I want to send received images from a user to Google Cloud Storage using server actions, but I get this error: Error: Only plain objects, and a few built-ins, can be passed to Server Actions. Classes or null prototypes are not supported. I read a bit about this problem on the internet and realised that it is due to the fact that next js cannot work with FileList or File. I'm using React Hook Form for the form. Is there a solution to this problem?

onSubmit form:
![Code_vVdurMpyBa](https://github.com/user-attachments/assets/61b7e91e-bee8-43f1-a6e9-fc8e44c7193b)

uploadImage server action:
![Code_3GFrg3Xc33](https://github.com/user-attachments/assets/31833961-3012-4918-91c6-58962c97dceb)

**Comment by lyubk0** [comment, marked as answer]
I figured it out. I don't know how correct this method is, but it worked for me.

![Code_UPUD8wsH1D](https://github.com/user-attachments/assets/0e96f447-0f7a-4e3e-b1ab-84093356002a)
![Code_dvx9RwuK3F](https://github.com/user-attachments/assets/f59dcb65-a56d-4ad1-a768-9816522b77fc)

**Comment by Tommy1223454** [comment]
React Hook Form can absolutely help with the client-side UX, but I’d still keep the security/trust decision fully on the server side.

For file uploads, I usually think of the flow as:
client form -> server receives raw file -> inspect -> only then persist/promote

I built a Node.js OSS package for that inspection layer:
https://github.com/pompelmi/pompelmi

It’s useful when you want upload handling to be more than just “the form submitted successfully”.
